--- original/src/merge3.c	2010-09-18 23:35:29.000000000 -0400
+++ fossil/src/merge3.c	2010-08-26 21:41:03.000000000 -0400
@@ -34,6 +34,8 @@
 #define ISDEBUG 0
 #endif
 
+int csv_merge(Blob *pPivot, Blob *pV1, Blob *pV2, Blob *pOut);
+
 /* The minimum of two integers */
 #define min(A,B)  (A<B?A:B)
 
@@ -159,6 +161,7 @@
   int nCpy, nDel, nIns;  /* Number of lines to copy, delete, or insert */
   int limit1, limit2;    /* Sizes of aC1[] and aC2[] */
   int nConflict = 0;     /* Number of merge conflicts seen so far */
+  int sub;
   static const char zBegin[] = ">>>>>>> BEGIN MERGE CONFLICT\n";
   static const char zMid[]   = "============================\n";
   static const char zEnd[]   = "<<<<<<< END MERGE CONFLICT\n";
@@ -266,6 +269,12 @@
         sz++;
       }
       DEBUG( printf("CONFLICT %d\n", sz); )
+      sub = csv_merge(pPivot, pV1, pV2, pOut);
+      if (sub==1 || sub==-1) {
+        free(aC1);
+        free(aC2);
+        return (sub==1)?0:-1;
+      }
       blob_appendf(pOut, zBegin);
       i1 = output_one_side(pOut, pV1, aC1, i1, sz);
       blob_appendf(pOut, zMid);
